tenant/pkg/controller/tenantnamespace/tenantnamespace_controller.go

```diff
@@ -167,45 +170,96 @@ func (r *ReconcileTenantNamespace) Reconcile(request reconcile.Request) (reconci
-			found := false
+			isOwner := false
 			for _, ownerRef := range each.OwnerReferences {
 				if ownerRef == expectedOwnerRef {
-					found = true
+					isOwner = true
 					break
 				} else if ownerRef.APIVersion == expectedOwnerRef.APIVersion && ownerRef.Kind == expectedOwnerRef.Kind {
 					// The namespace is owned by another TenantNamespace CR, fail the reconcile
 					err = fmt.Errorf("Namespace %v is owned by another %v TenantNamespace CR", each.Name, ownerRef)
 					return reconcile.Result{}, err
 				}
 			}
-			if !found {
+			if !isOwner {
 				log.Info("Namespace has been created without TenantNamespace owner", "namespace", each.Name)
 				// Obtain namespace ownership by setting ownerReference, and add annotation
 				err = r.updateNamespace(&each, &instance.Namespace, &expectedOwnerRef)
 			}
-			return reconcile.Result{}, err
+			break
 		}
 	}
 
 	// In case a new namespace needs to be created
-	tenantNs := &corev1.Namespace{
-		TypeMeta: metav1.TypeMeta{
-			APIVersion: corev1.SchemeGroupVersion.String(),
-			Kind:       "Namespace",
-		},
-		ObjectMeta: metav1.ObjectMeta{
-			Name: tenantNsName,
-			Annotations: map[string]string{
-				TenantAdminNamespaceAnnotation: instance.Namespace,
+	if !found {
+		tenantNs := &corev1.Namespace{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: corev1.SchemeGroupVersion.String(),
+				Kind:       "Namespace",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: tenantNsName,
+				Annotations: map[string]string{
+					TenantAdminNamespaceAnnotation: instance.Namespace,
+				},
+				OwnerReferences: []metav1.OwnerReference{expectedOwnerRef},
 			},
-			OwnerReferences: []metav1.OwnerReference{expectedOwnerRef},
-		},
+		}
+		if err = r.Client.Create(context.TODO(), tenantNs); err != nil {
+			return reconcile.Result{}, err
+		}
 	}
-	if err = r.Client.Create(context.TODO(), tenantNs); err != nil {
+
+	// Update tenant clusterrule to allow tenant admins to access the tenant namespace.
+	cr := &rbacv1.ClusterRole{}
+	if err = r.Get(context.TODO(), types.NamespacedName{Name: fmt.Sprintf("%s-tenant-admin-role", tenantName)}, cr); err != nil {
 		return reconcile.Result{}, err
 	}
-	return reconcile.Result{}, nil
+	cr = cr.DeepCopy()
+	foundNsRule := false
+	needUpdate := true
+	for i, each := range cr.Rules {
+		for _, resource := range each.Resources {
+			if resource == "namespaces" {
+				foundNsRule = true
+				break
+			}
+		}
+		if foundNsRule {
+			for _, resourceName := range each.ResourceNames {
+				if resourceName == tenantNsName {
+					needUpdate = false
+					break
+				}
+			}
+			if needUpdate {
+				cr.Rules[i].ResourceNames = append(cr.Rules[i].ResourceNames, tenantNsName)
+			}
+			break
+		}
+	}
+	if !foundNsRule {
+		err = fmt.Errorf("Cluster Role %s-tenant-admin-role does not have rules for namespaces.", tenantName)
+		return reconcile.Result{}, err
+	}
+	if needUpdate {
+		crClone := cr.DeepCopy()
+		err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
```

Comment:
nit: I’d prefer specifically checking this err in the block to prevent the inevitable someone missing it when adding more logic later.

Reply:
done